src/OpenTelemetry.Extensions.Hosting/Implementation/EventSourceEventFormatting.cs

```diff
@@ -0,0 +1,93 @@
+// <copyright file="EventSourceEventFormatting.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics.Tracing;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+
+namespace OpenTelemetry.Internal
+{
+    internal static class EventSourceEventFormatting
+    {
+        private static readonly StringBuilder Builder = new StringBuilder();
+
+        public static string Format(EventWrittenEventArgs eventData)
+        {
+            var payloadArray = eventData.Payload?.ToArray() ?? Array.Empty<object>();
+
+            ProcessPayloadArray(payloadArray);
+
+            if (eventData.Message != null)
+            {
+                try
+                {
+                    return string.Format(CultureInfo.InvariantCulture, eventData.Message, payloadArray);
+                }
+                catch (FormatException)
+                {
+                }
+            }
+
+            var stringBuilder = Builder;
```

Comment:
Will Format ever be called by more than one thread? Builder would get way corrupt in that case 😄

Reply:
Yeah, that could very well be an issue. I originally changed it to use a shared StringBuilder based on feedback from @reyang. I think what would be ideal is to make use of an ObjectPool, but we would need to bring in a package or do a source import to have access to an ObjectPool. It does seem like something that could be useful throughout this library.

Reply:
I just pushed a change that adds a simple `StringBuilderPool`.